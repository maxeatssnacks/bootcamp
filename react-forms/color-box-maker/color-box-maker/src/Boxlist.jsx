import React, { useState } from 'react';
import { v4 as uuid } from 'uuid';
import NewBoxForm from './NewBoxForm';
import Box from './Box';

const Boxlist = () => {
    const INITIAL_STATE = [];
    const [boxes, setBoxes] = useState(INITIAL_STATE);

    const createBox = (newBox) => {
        setBoxes(boxes => [...boxes, { ...newBox, id: uuid() }])
    }
    return (
        <div>
            <h3>Box-mania!</h3>
            <NewBoxForm createBox={createBox} />
            <div>
                {boxes.map(({ id, color, width, height }) => (
                    <Box key={id} id={id} color={color} width={width} height={height} />
                ))}
            </div>
        </div>
    )
}

export default Boxlist;