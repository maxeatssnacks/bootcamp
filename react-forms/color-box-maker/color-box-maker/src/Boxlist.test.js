import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Boxlist from './Boxlist';

describe('Boxlist', () => {
    // Smoke test
    test('renders without crashing', () => {
        render(<Boxlist />);
    });

    // Snapshot test
    test('matches snapshot', () => {
        const { asFragment } = render(<Boxlist />);
        expect(asFragment()).toMatchSnapshot();
    });

    // Critical business logic test
    test('can add a new box', () => {
        render(<Boxlist />);

        // Find form inputs and submit button
        const colorInput = screen.getByLabelText('Color:');
        const widthInput = screen.getByLabelText('Width:');
        const heightInput = screen.getByLabelText('Height:');
        const submitButton = screen.getByText('Create a new box!');

        // Fill out the form
        fireEvent.change(colorInput, { target: { value: '#FF0000' } });
        fireEvent.change(widthInput, { target: { value: '100' } });
        fireEvent.change(heightInput, { target: { value: '100' } });

        // Submit the form
        fireEvent.click(submitButton);

        // Check if a new box is added to the DOM
        const boxes = screen.getAllByRole('generic').filter(element => element.id);
        expect(boxes.length).toBe(1);
        expect(boxes[0]).toHaveStyle('background-color: #FF0000');
        expect(boxes[0]).toHaveStyle('width: 100px');
        expect(boxes[0]).toHaveStyle('height: 100px');
    });
});