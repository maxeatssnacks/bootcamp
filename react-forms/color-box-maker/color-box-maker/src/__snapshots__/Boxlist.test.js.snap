// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Boxlist matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Box-mania!
    </h3>
    <form>
      <label
        for="color"
      >
        Color: 
      </label>
      <input
        id="color"
        name="color"
        type="color"
        value="#ffffff"
      />
      <label
        for="width"
      >
        Width: 
      </label>
      <input
        id="width"
        name="width"
        type="number"
        value="100"
      />
      <label
        for="height"
      >
        Height: 
      </label>
      <input
        id="height"
        name="height"
        type="number"
        value="100"
      />
      <button
        type="submit"
      >
        Create a new box!
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
